src/vdaf/mastic.rs

```diff
@@ -0,0 +1,547 @@
+        mastic: &Mastic<T, P, SEED_SIZE>,
+        bytes: &mut Cursor<&[u8]>,
+    ) -> Result<Self, CodecError> {
+        MasticPublicShare::<VidpfWeight<T::Field>>::decode_with_param(
+            &(mastic.bits, mastic.vidpf.weight_parameter),
+            bytes,
+        )
+    }
+}
+
+/// Mastic input share
+///
+/// Message sent by the [`Client`] to each Aggregator during the Sharding phase.
+#[derive(Clone, Debug)]
+pub struct MasticInputShare<F: FieldElement, const SEED_SIZE: usize> {
+    /// VIDPF key share.
+    vidpf_key: VidpfKey,
+
+    /// The proof share.
+    proof_share: SzkProofShare<F, SEED_SIZE>,
+}
+
+impl<F: FieldElement, const SEED_SIZE: usize> Encode for MasticInputShare<F, SEED_SIZE> {
+    fn encode(&self, bytes: &mut Vec<u8>) -> Result<(), CodecError> {
+        bytes.extend_from_slice(&self.vidpf_key.value[..]);
+        self.proof_share.encode(bytes)?;
+        Ok(())
+    }
+
+    fn encoded_len(&self) -> Option<usize> {
+        Some(16 + self.proof_share.encoded_len()?)
+    }
+}
+
+impl<'a, T, P, const SEED_SIZE: usize> ParameterizedDecode<(&'a Mastic<T, P, SEED_SIZE>, usize)>
+    for MasticInputShare<T::Field, SEED_SIZE>
+where
+    T: Type,
+    P: Xof<SEED_SIZE>,
+{
+    fn decode_with_param(
+        (mastic, agg_id): &(&'a Mastic<T, P, SEED_SIZE>, usize),
+        bytes: &mut Cursor<&[u8]>,
+    ) -> Result<Self, CodecError> {
+        if *agg_id > 1 {
+            return Err(CodecError::UnexpectedValue);
+        }
+        let mut value = [0; 16];
+        bytes.read_exact(&mut value)?;
+        let vidpf_key = VidpfKey::new(
+            if *agg_id == 0 {
+                VidpfServerId::S0
+            } else {
+                VidpfServerId::S1
+            },
+            value,
+        );
+
+        let proof_share = SzkProofShare::<T::Field, SEED_SIZE>::decode_with_param(
+            &(
+                *agg_id == 0,
+                mastic.szk.typ.proof_len(),
+                mastic.szk.typ.joint_rand_len() != 0,
+            ),
+            bytes,
+        )?;
+        Ok(Self {
+            vidpf_key,
+            proof_share,
+        })
+    }
+}
+
+#[cfg(test)]
+impl<F: FieldElement, const SEED_SIZE: usize> PartialEq for MasticInputShare<F, SEED_SIZE> {
+    fn eq(&self, other: &MasticInputShare<F, SEED_SIZE>) -> bool {
+        self.ct_eq(other).into()
+    }
+}
+
+impl<F: FieldElement, const SEED_SIZE: usize> ConstantTimeEq for MasticInputShare<F, SEED_SIZE> {
+    fn ct_eq(&self, other: &MasticInputShare<F, SEED_SIZE>) -> Choice {
+        self.vidpf_key
+            .ct_eq(&other.vidpf_key)
+            .bitand(self.proof_share.ct_eq(&other.proof_share))
+    }
+
+    fn ct_ne(&self, other: &MasticInputShare<F, SEED_SIZE>) -> Choice {
+        self.ct_eq(other).not()
+    }
+}
+
+/// Mastic output share
+///
+/// Contains a vector of VIDPF outputs: one for each prefix.
+pub type MasticOutputShare<V> = OutputShare<V>;
+
+/// Vector of VidpfWeights to be aggregated by Mastic aggregators
```

Comment:
This comment needs to be updated. Also, start with "Mastic aggregate share." like you've done for other types.